### PR TITLE
Support multiple stop tokens

### PR DESCRIPTION
**Pull request by lucasavila00**

### CODE REVIEW

**Comment by lucasavila00** [comment]
@EricLBuehler how should I profile this change?

Should I run on CPU to avoid CUDA's async API? The only place I feel I can grasp the sampling duration is on the visual profiler :thinking:

**Comment by EricLBuehler** [comment]
@lucasavila00, actually, I was thinking of something a lot simpler: comparing the completion T/s on master to this branch. Ideally the cost would be opt-in, and when it is used it would not be too costly.

It looks like your usage of `decoded_tokens` is actually incorrect: it stores the decoded tokens for streaming (the naming is not the best, we should probably change it). Perhaps you could cache the last generated token in `Sequence` (entrypt: `add_token`) or pass it to `is_done` (extract in `sample_seqs`)?

**Comment by EricLBuehler** [comment]
I think this is still unfortunately incorrect, as getting the delta is only for streaming sequences. Can you try to get the `.bytes` field of `Logprobs` (returned by the sampler) and pass that to the `is_done` function instead?

**Comment by lucasavila00** [comment]
> I think this is still unfortunately incorrect, as getting the delta is only for streaming sequences. Can you try to get the .bytes field of Logprobs (returned by the sampler) and pass that to the is_done function instead?

Sure, but won't that have just the last token's bytes?

The check must be done on the whole string, no? For instance, I can set the stop to "a word and spaces" and to know if we have generated this text we can't really optimize it, no?


I was able to reproduce the slowdown even when not being used, I'm trying to figure that out first.

**Comment by EricLBuehler** [comment]
Hmm, that is a good point. Perhaps we could keep a String containing the entire decoded sequence, and then check against that? I think that would be faster than decoding all the tokens every pass. I also found the slowdown when not being used, that was very confusing.

**Comment by lucasavila00** [comment]
@EricLBuehler it was caused by cloning the tokenizer. I moved the tokenizer to an Arc and it fixed the issue.

**Comment by lucasavila00** [comment]
>Ah, great that you found that! Is there a performance regression when it checks against the stop seqs?

I'm starting on it.

BTW the ARC will probably fix the streaming issues #99 as it also cloned the tokenizer here. I assumed that it was cheap because of it.

**Comment by lucasavila00** [comment]
> Is there a performance regression when it checks against the stop seqs?

When used, on https://github.com/EricLBuehler/mistral.rs/pull/116/commits/d3d15cb107e5445396b4b8f92d6acd37566f7f0f there is a performance regression.

I'm getting 73 for this PR and 77 for master (warm run, prefilled cache)

Let me try avoiding re-tokenizing everything all the time now.

**Comment by lucasavila00** [comment]
I cannot figure out how to re-build the text from the tokens iteratively.

If I get the values directly from tokenizer using `id_to_token` special characters appear in the text

```
2024-04-12T03:03:41.495875Z  INFO mistralrs_server: Serving on http://0.0.0.0:1234.
[mistralrs-core/src/sequence.rs:269:17] &self.completion_text = "▁Once"
[mistralrs-core/src/sequence.rs:269:17] &self.completion_text = "▁Once▁upon"
[mistralrs-core/src/sequence.rs:269:17] &self.completion_text = "▁Once▁upon▁a"
```

or if I use `tokenizer.decode` then text has no spaces.
```
2024-04-12T02:50:17.690452Z  INFO mistralrs_server: Serving on http://0.0.0.0:1234.
[mistralrs-core/src/sequence.rs:269:17] &self.completion_text = "Once"
[mistralrs-core/src/sequence.rs:269:17] &self.completion_text = "Onceupon"
```

I organized the code and left it still re-tokenizing everything all the time.

**Comment by lucasavila00** [comment]
I think the TokTrie might have utilities that help with this. Let me give it a shot.

**Comment by lucasavila00** [comment]
After https://github.com/EricLBuehler/mistral.rs/pull/116/commits/404af701f0cef06d9850901db62eebc7c8e1117c

I measure 76t/s when enabled, 77t/s when disabled.

File: mistralrs-core/src/sequence.rs [hunk 1]
```diff
@@ -243,21 +249,45 @@ impl Sequence {
...
+    fn is_sub<T: PartialEq>(mut haystack: &[T], needle: &[T]) -> bool {
+        if needle.is_empty() {
+            return true;
+        }
+        while !haystack.is_empty() {
+            if haystack.starts_with(needle) {
+                return true;
+            }
+            haystack = &haystack[1..];
+        }
```
Comment (lucasavila00) [inline mistralrs-core/src/sequence.rs:261]: Maybe it's worth it to use a more optimized version of this is_sub?

eg: https://docs.rs/galil-seiferas/latest/galil_seiferas/ from https://stackoverflow.com/a/47045355

Comment (lucasavila00) [reply mistralrs-core/src/sequence.rs]: I added it...

Comment (lucasavila00) [reply mistralrs-core/src/sequence.rs]: I'm not sure if this is the best library...

**Review by EricLBuehler — APPROVED** [review]

**Comment by EricLBuehler** [comment]
Thank you!

**Pull request merged**